proposals/11-21-krm-fn-k8s-backend.md

```diff
@@ -0,0 +1,218 @@
+# Implementation of k8s backend to run krm functions
+
+**Authors**: alexey.odinokov.82@gmail.com
+
+**Reviewers**: <!-- List at least one Kustomize approver (https://github.com/kubernetes-sigs/kustomize/blob/master/OWNERS#L2) -->
+
+**Status**: implementable
+
+## Summary
+
+Currently Kustomize supports plugin-extensions via containers that use the certain API:
+see [KRM-functions](https://github.com/kubernetes-sigs/kustomize/blob/master/cmd/config/docs/api-conventions/functions-spec.md).
+This is very convenient because of many reasons, e.g. it doesn't require prior plugin-installation, and in the current
+implementation it requires only `docker` installed and running: `kustomize` calls `docker` client
+to run the specified docker-container. Unfortunately, in cases when `kustomize` has to be ran
+inside another container, e.g. inside a pod like in case of k8s it requires to use docker-in-docker or
+mount docker-socket, which is not always possible. And in cases when it's possible this in its turn
+creates security concerns, because docker-in-docker requires container to be ran in privileged mode
+and shared docker-socket also introduces issues: it allows to control containers created by others. 
+In addition, its `--mount` flag won't work correctly, because source path is treated like a path in the
+host environment, not inside the container. And that allows to expose some files from host to the container.
+
+This proposal aims to avoid that concerns by introducing the alternative approach where krm-functions
+are ran as a pod in k8s cluster. That approach allows to use k8s-native mechanisms to control that pod.
+Moreover, since `kustomize` is included into `kubectl` that approach removes a mandatory dependency
+on `docker` and makes `kubectl` self-sufficient to run krm-function based plugins.
+
+## Motivation
+
+1. https://github.com/GoogleContainerTools/kpt/issues/2158
+1. https://github.com/kubernetes-sigs/kustomize/pull/4260
+
+
+**Goals:**
```

Comment:
Another goal we should try to achieve here is to design the schema for supporting it in kustomization.yaml.
In kustomize, users may want to express the intend declaratively.
https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/2953-kustomize-plugin-graduation may be revelant.

Reply:
That is a great doc. I intentionally didn't want to change anything in the current kustomization.yaml and as you see I selected the approach to use the same `container` spec that we already have - because I wanted users not to change anything in their kustomizations by just allowing them to select where that container should run: locally or in k8s. 
That was the reason I put 'Change Krm-functions spec' in the the non-goals section. My goal was to keep that spec untouched, so it shouldn't affect the schema.

Reply:
I've updated the goals based on that comment